src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/AvoidOutParameters.cs

```diff
@@ -0,0 +1,70 @@
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.AvoidOutParametersDescription), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources));
+
+        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(RuleId,
+                                                                             s_localizableTitle,
+                                                                             s_localizableMessage,
+                                                                             DiagnosticCategory.Design,
+                                                                             DiagnosticHelpers.DefaultDiagnosticSeverity,
+                                                                             isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+                                                                             description: s_localizableDescription,
+                                                                             helpLinkUri: "https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1021",
+                                                                             customTags: FxCopWellKnownDiagnosticTags.PortedFxCopRule);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterCompilationStartAction(csac =>
+            {
+                var outAttributeType = csac.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeInteropServicesOutAttribute);
+
+                csac.RegisterSymbolAction(sac =>
+                {
+                    var methodSymbol = (IMethodSymbol)sac.Symbol;
+
+                    var numberOfOutParams = methodSymbol.Parameters.Count(p => IsOutParameter(p, outAttributeType));
+
+                    if (numberOfOutParams >= 1 && !IsTryPatternMethod(methodSymbol, outAttributeType, numberOfOutParams))
+                    {
+                        sac.ReportDiagnostic(methodSymbol.CreateDiagnostic(Rule));
+                    }
+                }, SymbolKind.Method);
+            });
+        }
+
+        private static bool IsOutParameter(IParameterSymbol parameterSymbol, INamedTypeSymbol? outAttributeSymbol) =>
+            parameterSymbol.RefKind == RefKind.Out ||
+            // Handle VB.NET special case for out parameters
+            (parameterSymbol.RefKind == RefKind.Ref && parameterSymbol.HasAttribute(outAttributeSymbol));
```

Comment:
I couldn't find any better way for handling this case. It seems that there is an `IsOut` property on the class itself but that's not available through the interface.

Reply:
If there isn't any better way, I think that it might be worth to have such method directly available in Roslyn. WDYT?